public/app/plugins/panel/geomap/layers/data/heatMap.tsx

```diff
@@ -0,0 +1,243 @@
+  },
+  blur: 15,
+  radius: 5,
+};
+
+/**
+ * Map layer configuration for heatmap overlay
+ */
+export const heatmapLayer: MapLayerRegistryItem<HeatmapConfig> = {
+  id: 'heatmap',
+  name: 'Heatmap',
+  description: 'visualizes a heatmap of the data',
+  isBaseMap: false,
+
+  /**
+   * Function that configures transformation and returns a transformer
+   * @param options
+   */
+  create: (map: Map, options: MapLayerConfig<HeatmapConfig>, theme: GrafanaTheme2): MapLayerHandler => {
+    const config = { ...defaultOptions, ...options.config };
+    const vectorSource = new source.Vector();
+
+    // Create a new Heatmap layer
+    // Weight function takes a feature as attribute and returns a normalized weight value
+    const vectorLayer = new layer.Heatmap({
+      source: vectorSource,
+      blur: config.blur,
+      radius: config.radius,
+      weight: function (feature) {
+        var weight = feature.get('value');
+        return weight;
+      },
+    });
+
+    return {
+      init: () => vectorLayer,
+      update: (data: PanelData) => {
+        const frame = data.series[0];
+
+        // Remove previous data before updating
+        const features = vectorLayer.getSource().getFeatures();
+        features.forEach((feature) => {
+          vectorLayer.getSource().removeFeature(feature);
+        });
+
+        // Get data points (latitude and longitude coordinates)
+        const points = dataFrameToPoints(frame, config.fieldMapping, config.queryFormat);
+        // Get the field of data values
+        const field = frame.fields.find(field => field.name === config.fieldMapping.metricField);
+        // Return early if metric field is not matched
+        if (field === undefined) {
+          return;
+        };
+
+        // Retrieve the min, max and range of data values
+        const calcs = reduceField({
+          field: field,
+          reducers: [
+            ReducerID.min,
+            ReducerID.range,
+          ]
+        });
+        // Map each data value into new points
+        points.map((point, i) => {
+          const cluster = new Feature({
+              geometry: point,
+              value: normalize(calcs, field.values.get(i)),
+          });
+          vectorSource.addFeature(cluster);
+        });
+        vectorLayer.setSource(vectorSource);
+
+        // Set gradient of heatmap
+        const colorMode = getFieldColorModeForField(field);
+        if (colorMode.isContinuous && colorMode.getColors) {
+          // getColors return an array of color string from the color scheme chosen
+          const colors = colorMode.getColors(theme);
+          vectorLayer.setGradient(colors);
+        } else {
+          // Set the gradient back to default if threshold or single color is chosen
+          vectorLayer.setGradient(['#00f', '#0ff', '#0f0', '#ff0', '#f00']);
+        }
+      },
+    };
+  },
+  // Heatmap overlay options
+  registerOptionsUI: (builder) => {
+    builder
+      .addSelect({
```

Comment:
With https://github.com/grafana/grafana/pull/36768, you can now remove all the queryFormat+spatial type configs

Reply:
Updated!